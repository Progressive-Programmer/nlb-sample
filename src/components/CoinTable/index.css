.cointable-component {
    display: grid;
    justify-items: center;
    margin-right: auto;
    margin-left: auto;
    margin-top: 20px;
}