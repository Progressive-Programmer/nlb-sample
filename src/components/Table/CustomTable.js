import React, { useEffect, useState } from "react";
import Paper from "@mui/material/Paper";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TablePagination from "@mui/material/TablePagination";
import TableRow from "@mui/material/TableRow";
import Avatar from "@mui/material/Avatar";
import ListItemAvatar from "@mui/material/ListItemAvatar";
import Loader from "../Backdrop/Loader";
import CoinCard from "../CoinCard/CoinCard";

const CustomTable = (props) => {
  const { isLoading, tableData, columns, coinData, onPageChange, onRowClick } =
    props;
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [controller, setController] = useState({
    page: 0,
    rowsPerPage: 10,
  });

  useEffect(() => {
    // fetch table data on first render and
    // then on every page change
    getData();
  }, [controller]);

  const getData = async () => {
    // material ui default page count starts with zero
    // and the first page for api is one
    await onPageChange(controller.page + 1, controller.rowsPerPage);
  };

  const handleChangePage = (event, newPage) => {
    setController({
      ...controller,
      page: newPage,
    });
  };

  const handleChangeRowsPerPage = (event) => {
    setController({
      ...controller,
      rowsPerPage: parseInt(event.target.value, 10),
      page: 0,
    });
  };

  const handleClickRow = (row) => {
    onRowClick(row.id);
    setIsModalOpen(true);
  };

  return (
    <Paper sx={{ width: "90%", overflow: "hidden" }}>
      {isLoading ? <Loader isOpen={isLoading} /> : null}
      <TableContainer sx={{ maxHeight: 650 }}>
        <Table stickyHeader aria-label="sticky table" size="small">
          <TableHead>
            <TableRow>
              {columns &&
                columns.map(
                  (
                    column // set custom column names
                  ) => (
                    <TableCell
                      key={column.id}
                      align={column.align}
                      style={{ minWidth: column.minWidth }}
                    >
                      {column.label}
                    </TableCell>
                  )
                )}
            </TableRow>
          </TableHead>
          <TableBody>
            {tableData
              ? tableData // maping data to each custom row
                  .map((row, i) => {
                    return (
                      <TableRow
                        hover
                        role="checkbox"
                        tabIndex={-1}
                        key={`${row.code}-${i}`}
                        style={{ cursor: "pointer" }}
                        onClick={() => handleClickRow(row)}
                      >
                        {columns &&
                          columns.map((column) => {
                            const value = row[column.id];
                            if (column.id === "image") {
                              // table cell spearation for image type and text
                              return (
                                <TableCell key={column.id} align={column.align}>
                                  <ListItemAvatar>
                                    <Avatar alt={row.name} src={value} />
                                  </ListItemAvatar>
                                </TableCell>
                              );
                            } else {
                              return (
                                <TableCell key={column.id} align={column.align}>
                                  {column.format && typeof value === "number"
                                    ? column.format(value)
                                    : value}
                                </TableCell>
                              );
                            }
                          })}
                      </TableRow>
                    );
                  })
              : "No Data Available"}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={100}
        rowsPerPage={controller.rowsPerPage}
        page={controller.page}
        onPageChange={handleChangePage}
        onRowsPerPageChange={handleChangeRowsPerPage}
      />
      {!isLoading && isModalOpen && (
        <CoinCard
          isLoading={isLoading}
          isOpen={isModalOpen}
          data={coinData ? coinData : {}}
          setIsModalOpen={setIsModalOpen}
        />
      )}
    </Paper>
  );
};

export default CustomTable;
