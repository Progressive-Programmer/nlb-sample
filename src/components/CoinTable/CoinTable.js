import React from "react";
import CustomTable from "../Table/CustomTable";
import "./index.css";

const columns = [
  { id: "image", label: "Coin", minWidth: 100 },
  { id: "name", label: "Name", minWidth: 100 },
  {
    id: "symbol",
    label: "Symbol",
    minWidth: 100,
  },
  {
    id: "current_price",
    label: "Current Price",
    minWidth: 170,
    align: "right",
    format: (value) => `$${value}`
  },
  {
    id: "high_24h",
    label: "High 24 Hr Price",
    minWidth: 170,
    align: "right",
    format: (value) => `$${value}`
  },
  {
    id: "low_24h",
    label: "Low 24 Hr Price",
    minWidth: 170,
    align: "right",
    format: (value) => `$${value}`
  },
];

const CoinTable = (props) => {
  const {
    isLoading,
    hasError,
    currentResponseId,
    currentPage,
    markets,
    coinData,
    getCoinMarketsBypage,
    getCoinDateByCoinId,
  } = props;

  const handlePageChange = async (pageNo, perPage) => {
    try {
      await getCoinMarketsBypage({ pageNo, perPage });
    } catch (error) {
      console.log(error)
    }
  };

  const handleRowClick = async (coinId) => {
    try {
      await getCoinDateByCoinId({ coinId })
    } catch (error) {
     console.log(error) 
    }
  }
  return (
    <div className="cointable-component">
      <CustomTable
        isLoading={isLoading}
        currentPage={currentPage}
        tableData={markets}
        coinData={coinData}
        columns={columns}
        onPageChange={handlePageChange}
        onRowClick={handleRowClick}
      />
    </div>
  );
};

export default CoinTable;
