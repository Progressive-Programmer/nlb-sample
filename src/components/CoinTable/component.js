import React, { useEffect } from "react";
import CustomTable from "../Table/component";
import "./index.css";

const columns = [
  { id: "image", label: "Coin", minWidth: 100 },
  { id: "name", label: "Name", minWidth: 100 },
  {
    id: "symbol",
    label: "Symbol",
    minWidth: 170,
    align: "right",
  },
  {
    id: "current_price",
    label: "Current Price",
    minWidth: 170,
    align: "right",
  },
  {
    id: "high_24h",
    label: "High 24 Hr Price",
    minWidth: 170,
    align: "right",
  },
  {
    id: "low_24h",
    label: "Low 24 Hr Price",
    minWidth: 170,
    align: "right",
  },
];

const CoinTable = (props) => {
  const {
    isLoading,
    hasError,
    currentResponseId,
    currentPage,
    markets,
    coinData,
    getCoinMarketsBypage,
    getCoinDateByCoinId,
  } = props;

  const onPageChange = async (pageNo, perPage) => {
    await getCoinMarketsBypage({ pageNo, perPage });
  };
  return (
    <div className="cointable-component">
      <CustomTable
        isLoading={isLoading}
        hasError={hasError}
        currentPage={currentPage}
        tableData={markets}
        columns={columns}
        onPageChange={onPageChange}
      />
    </div>
  );
};

export default CoinTable;
