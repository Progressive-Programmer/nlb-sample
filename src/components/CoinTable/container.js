import { connect } from "react-redux";
import { getCoinDataById, getCoinMarkets } from "../../redux/slices/coinSlice";
import CoinTable from "./component";

const mapStateToProps = (state, ownPros) => {
  let coins = state.coins
  return {
    isLoading: coins.loading,
    hasError: coins.error,
    currentResponseId: coins.currentResponseId,
    currentPage: coins.currentPage,
    markets: coins.markets,
    coinData: coins.coinData,
  };
};

const mapDispatchToProps = (dispatch, ownPros) => {
  return {
    getCoinMarketsBypage: (pageNo) => dispatch(getCoinMarkets(pageNo)),
    getCoinDateByCoinId: (coinId) => dispatch(getCoinDataById(coinId)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(CoinTable);
