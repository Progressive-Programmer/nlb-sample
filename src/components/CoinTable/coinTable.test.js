import React from 'react';
import { shallow } from 'enzyme';
import CoinTable from './CoinTable';
import CustomTable from '../Table/CustomTable';

const setUp = (props = {}) => {
  const wrapper = shallow(<CoinTable {...props} />);
  return wrapper;
};

describe('CoinTable', () => {
  describe('has props', () => {
    let wrapper;
    let props;
    let state;
    beforeEach(() => {
      props = {
        isLoading: false,
        currentPage: 0,
        markets: [
          { id: 1, image: 'image', name: 'name', symbol: 'byt', current_price: 1 },
          { id: 2, image: 'image', name: 'name', symbol: 'byt', current_price: 1 },
          { id: 3, image: 'image', name: 'name', symbol: 'byt', current_price: 1 }
        ],
        coinData: {},
        getCoinMarketsBypage: jest.fn(),
        getCoinDateByCoinId: jest.fn(),
        columns: [
          { id: 'image', label: 'Coin', minWidth: 100 },
          { id: 'name', label: 'Name', minWidth: 100 },
          {
            id: 'symbol',
            label: 'Symbol',
            minWidth: 100
          },
          {
            id: 'current_price',
            label: 'Current Price',
            minWidth: 170,
            align: 'right',
            format: (value) => `$${value}`
          },
          {
            id: 'high_24h',
            label: 'High 24 Hr Price',
            minWidth: 170,
            align: 'right',
            format: (value) => `$${value}`
          },
          {
            id: 'low_24h',
            label: 'Low 24 Hr Price',
            minWidth: 170,
            align: 'right',
            format: (value) => `$${value}`
          }
        ]
      };
      wrapper = setUp(props, state);
    });

    afterEach(() => {
      jest.resetAllMocks();
    });
    it('should render Table Component', () => {
      const component = wrapper;
      expect(component.find('.cointable-component')).toHaveLength(1);
    });
    it('should call getCoinMarketsBypage when page is changed', async () => {
      const pageNumber = 0;
      const perPage = 10;
      const getCoinMarketsBypage = await props.getCoinMarketsBypage(pageNumber, perPage);
      const handleOnPageChange = jest.spyOn(props, 'getCoinMarketsBypage');
      const component = shallow(
        <CoinTable {...props} getCoinMarketsBypage={getCoinMarketsBypage} />
      );

      component.find(CustomTable).simulate('onPageChange');
      expect(handleOnPageChange).toHaveBeenCalledWith(pageNumber, perPage);
    });
    it('should catch error if getCoinMarketsBypage does crashes', async () => {
      const pageNumber = 0;
      const perPage = 10;
      try {
        await props.getCoinMarketsBypage(pageNumber, perPage);
      } catch (e) {
        expect(e).toThrowError();
      }
    });
  });
});
