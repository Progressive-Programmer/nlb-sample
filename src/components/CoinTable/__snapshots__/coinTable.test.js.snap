// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinTable has props should render correctly 1`] = `
<div
  className="cointable-component"
>
  <CustomTable
    coinData={Object {}}
    columns={
      Array [
        Object {
          "id": "image",
          "label": "Coin",
          "minWidth": 100,
        },
        Object {
          "id": "name",
          "label": "Name",
          "minWidth": 100,
        },
        Object {
          "id": "symbol",
          "label": "Symbol",
          "minWidth": 100,
        },
        Object {
          "align": "right",
          "format": [Function],
          "id": "current_price",
          "label": "Current Price",
          "minWidth": 170,
        },
        Object {
          "align": "right",
          "format": [Function],
          "id": "high_24h",
          "label": "High 24 Hr Price",
          "minWidth": 170,
        },
        Object {
          "align": "right",
          "format": [Function],
          "id": "low_24h",
          "label": "Low 24 Hr Price",
          "minWidth": 170,
        },
      ]
    }
    currentPage={0}
    isLoading={false}
    onPageChange={[Function]}
    onRowClick={[Function]}
    tableData={
      Array [
        Object {
          "current_price": 1,
          "id": 1,
          "image": "image",
          "name": "name",
          "symbol": "byt",
        },
        Object {
          "current_price": 1,
          "id": 2,
          "image": "image",
          "name": "name",
          "symbol": "byt",
        },
        Object {
          "current_price": 1,
          "id": 3,
          "image": "image",
          "name": "name",
          "symbol": "byt",
        },
      ]
    }
  />
</div>
`;

exports[`CustomTable has props should render correctly 1`] = `
<div
  className="cointable-component"
>
  <CustomTable
    coinData={Object {}}
    columns={
      Array [
        Object {
          "id": "image",
          "label": "Coin",
          "minWidth": 100,
        },
        Object {
          "id": "name",
          "label": "Name",
          "minWidth": 100,
        },
        Object {
          "id": "symbol",
          "label": "Symbol",
          "minWidth": 100,
        },
        Object {
          "align": "right",
          "format": [Function],
          "id": "current_price",
          "label": "Current Price",
          "minWidth": 170,
        },
        Object {
          "align": "right",
          "format": [Function],
          "id": "high_24h",
          "label": "High 24 Hr Price",
          "minWidth": 170,
        },
        Object {
          "align": "right",
          "format": [Function],
          "id": "low_24h",
          "label": "Low 24 Hr Price",
          "minWidth": 170,
        },
      ]
    }
    currentPage={0}
    isLoading={false}
    onPageChange={[Function]}
    onRowClick={[Function]}
    tableData={
      Array [
        Object {
          "current_price": 1,
          "id": 1,
          "image": "image",
          "name": "name",
          "symbol": "byt",
        },
        Object {
          "current_price": 1,
          "id": 2,
          "image": "image",
          "name": "name",
          "symbol": "byt",
        },
        Object {
          "current_price": 1,
          "id": 3,
          "image": "image",
          "name": "name",
          "symbol": "byt",
        },
      ]
    }
  />
</div>
`;
